### PR TITLE
bugfix and improvements

### PR DESCRIPTION
**Pull request by ycheng627**
Hi Reviewers,

These are a collection of CLs that change/fixes various behaviors for the reference app. PTAL!

### CODE REVIEW

File: diagnostics-app/src/app/events/events-dashboard/events-dashboard.component.ts [hunk 1]
```diff
@@ -45,6 +45,7 @@ export class EventsDashboardComponent implements OnInit {
       statusInfo = statusInfo as EventSupportStatusInfo;
```
Comment (BaiiYuan) [inline diagnostics-app/src/app/events/events-dashboard/events-dashboard.component.ts:45]: Should we update `_visibleCategories` in this file to use `VISIBLE_EVENT_CARDS`?

File: diagnostics-app/src/config/config.ts [hunk 2]
```diff
@@ -25,6 +25,24 @@ export const VISIBLE_TELEMETRY_CARDS: TelemetryInfoType[] = [
   TelemetryInfoType.TPM,
 ];
 
+export const VISIBLE_EVENT_CARDS: EventCategory[] = [
```
Comment (BaiiYuan) [inline diagnostics-app/src/config/config.ts:28]: Could you mention this in the commit message?

Maybe `Move visible event cards to customization config`

File: diagnostics-app/src/app/core/services/events.service.ts [hunk 3]
```diff
@@ -106,37 +102,44 @@ export class EventsService {
...
-  }
+  stopCapturingEvents:
+      (eventType: EventCategory) => Promise<EventsResponse> = eventType => {
+        const payload: EventsRequest = {
+          action: EventsAction.STOP_CAPTURING_EVENT,
+          eventType,
+        };
+        const request = this.constructEventsRequest(payload);
+        return this.sendRequest(request);
+      }
```
Comment (BaiiYuan) [inline diagnostics-app/src/app/core/services/events.service.ts:144]: Suggestion: It seems that this part is updated by formatter. (Ditto to the import headers)

Maybe we should have a unified way to format these typescript files.

I think we need to move this kind of change into another PR if we want to use different formatter. Otherwise this PR will have a lot of unnecessary changes, which make it hard to notice the logic change.

Thanks!

**Comment by BaiiYuan** [comment]
nit: Remove the prefix in the PR title and have a summary in comment#1.

File: diagnostics-app/src/app/events/events-dashboard/events-dashboard.component.ts (hunk 1 above)
Comment (ycheng627) [reply diagnostics-app/src/app/events/events-dashboard/events-dashboard.component.ts:45]: Done.

File: diagnostics-app/src/config/config.ts (hunk 2 above)
Comment (ycheng627) [reply diagnostics-app/src/config/config.ts:28]: Done.

File: diagnostics-app/src/app/core/services/events.service.ts (hunk 3 above)
Comment (ycheng627) [reply diagnostics-app/src/app/core/services/events.service.ts]: Done.

You are correct, I've opened up a bug b/308944438 and outlined my approach there.

I'm planning to run formatter on all files after this last PR is merged.

**Comment by ycheng627** [comment]
Hi Byron,

Updated according to comments. PTAL again!

File: diagnostics-app/src/app/app.module.ts [hunk 4]
```diff
@@ -24,6 +24,13 @@ import { HeaderComponent } from './layout/header/header.component';
 import { SharedModule } from './shared/shared.module';
 import { SideNavComponent } from './layout/side-nav/side-nav.component';
 import { TelemetryModule } from './telemetry/telemetry.module';
+import { EventsService } from './core/services/events.service';
+
+const initializeEventService = (eventsService: EventsService): () => Promise<any> => {
+  return () => {
```
Comment (andy920262) [inline diagnostics-app/src/app/app.module.ts:30]: nit: `return () => eventsService.Init();`

File: diagnostics-app/src/app/core/services/events.service.ts [hunk 5]
```diff
@@ -74,30 +79,44 @@ export class EventsService {
...
+    })
+  }
 
-    } catch (err) {
-      console.error(ResponseErrorInfoMessage.FailedEventsServiceConstructor);
-    }
+  constructor() {
+    this.extensionId = environment.extensionId;
+    this.supportabilityCache = new Map<EventCategory, Promise<EventsResponse>>;
```
Comment (andy920262) [inline diagnostics-app/src/app/core/services/events.service.ts:110]: We can just initialize these in line 47:
```
private readonly extensionId = environment.extensionId;
private supportabilityCache = new Map<EventCategory, Promise<EventsResponse>>;
```

File: diagnostics-app/src/app/core/services/events.service.ts [hunk 6]
```diff
@@ -118,8 +137,13 @@ export class EventsService {
...
       eventType,
     };
+    if (this.supportabilityCache.has(eventType)) {
+      return this.supportabilityCache.get(eventType)!;
+    }
     const request = this.constructEventsRequest(payload);
-    return this.sendRequest(request);
+    const promise = this.sendRequest(request);
+    this.supportabilityCache.set(eventType, promise);
+    return promise;
```
Comment (andy920262) [inline diagnostics-app/src/app/core/services/events.service.ts:146]: nit: I think this is more clear
```
    if (!this.supportabilityCache.has(eventType)) {
         const request = this.constructEventsRequest(payload);
         const promise = this.sendRequest(request);
         this.supportabilityCache.set(eventType, promise);
    }
    return this.supportabilityCache.get(eventType)!;
```

File: diagnostics-app/src/app/diagnostics/diagnostics-card/diagnostics-card.component.ts [hunk 7]
```diff
@@ -113,7 +113,12 @@ export class DiagnosticsCardComponent implements OnInit, OnDestroy {
       const interval = (refreshIntervals.diagnostics.has(this.routine)) ?
         refreshIntervals.diagnostics.get(this.routine) : defaultDiagnosticsRefreshInterval;
       this._intervalId = window.setInterval(() => {
-        this.getRoutineStatus();
+        // In case there is a race condition where additional getRoutineStatus
+        // is run after the routine has finished, check that the interval ID is
+        // still set.
+        if (this.intervalId !== undefined) {
```
Comment (andy920262) [inline diagnostics-app/src/app/diagnostics/diagnostics-card/diagnostics-card.component.ts:119]: How about introduce a new state `STOPPING`?

Comment (ycheng627) [reply diagnostics-app/src/app/diagnostics/diagnostics-card/diagnostics-card.component.ts]: Done.

File: diagnostics-app/src/app/core/services/events.service.ts (hunk 6 above)
Comment (ycheng627) [reply diagnostics-app/src/app/core/services/events.service.ts]: Done.

File: diagnostics-app/src/app/core/services/events.service.ts (hunk 5 above)
Comment (ycheng627) [reply diagnostics-app/src/app/core/services/events.service.ts]: Done.

File: diagnostics-app/src/app/app.module.ts (hunk 4 above)
Comment (ycheng627) [reply diagnostics-app/src/app/app.module.ts]: Done.

**Comment by ycheng627** [comment]
Hi reviewers,

I've edit based on comments. PTAL!

**Review by BaiiYuan — COMMENTED** [review]
Overall LGTM

**Review by andy920262 — APPROVED** [review]
